### ISSUE TITLE
Plex webhook not working

### ISSUE THREAD

**Issue by mwckr** [IgnisDa/ryot#814]
The Plex webhook is not working for me.

Running the latest Ryot and Plex versions on Proxmox.
Both are on the local network on separate VMs. Ryot is also behind my reverse proxy (NGINX) and available via ryot.domain.org (domain not being my actual domain).

I created the Plex webhook via the Ryot integrations page. I tried both via my local-ip and my reverse-proxy-ip. Tried my full name from my profile and my E-Mail from my account.
The format is: https://ryot.domain.org/backend/webhooks/integrations/plex/<slug> --> I did paste this into my webhooks settings in Plex. Again, both with the domain or the local IP.

Nothing is updating in ryot. I have added a show which I'm currently watching on Plex to my "In Progress" shows, and nothing is happening. Multiple episodes watches, no progress in Ryot.

I don't know what i'm doing wrong or if this is a bug. Any help would be appreciated.

**Comment by IgnisDa** [comment]
Are you sure you copied the entire webhook? There should be a random slug at the end of the webhook url.

**Comment by mwckr** [comment]
> Are you sure you copied the entire webhook? There should be a random slug at the end of the webhook url.

Yes, sure. I just redacted the slug in my issue test above. 

I created the webhook, copy the whole thing (which is easily done because it selects the whole field with one click, which is nicely done), copy and paste it in Plex. I even checked it character by character. 

Nothing updates. 

Let me see if I'm expecting the right behavior:

I am watching Curb your Enthusiasm on Plex. Next up is episode 9 of season 2. So I added Curb via the TV shows search to my library and, by marking all episodes as seen so the progress is in sync with Plex, put it "in Progress" (I can see the tag on top). 

Now I watch episode 9 on Plex. What I expect should happen is that it now marks episode 9 as seen on Ryot. But that doesn't happen. 

The show is correctly matched on Plex. It's suggesting the same posters on both Plex and Ryot. 

Anything I'm missing?

**Comment by IgnisDa** [comment]
There must be some other log which is raised in Plex to indicate that the webhook request has failed. Can you find out the problem that way?

**Comment by mwckr** [comment]
> There must be some other log which is raised in Plex to indicate that the webhook request has failed. Can you find out the problem that way?

When should I check? When saving the webhook? When watching or ending media?

**Comment by IgnisDa** [comment]
The Plex logs etc. It must've somewhere were it logs the status code of sending webhooks. Honestly I'm not sure since I don't have Plex. But i do know that some logs are shown because the person who developed the webhook code showed it to me.

**Comment by mwckr** [comment]
I will check a couple of places and get back to you. Thanks.

**Comment by mwckr** [comment]
So in the browser console there was nothing to go on, but I downloaded the logs and found this in the "Plex Media Server.log" file:

First error:
May 05, 2024 10:55:30.127 [124506692152120] DEBUG - [Req#f53be/HCl#62a2] HTTP requesting POST http://192.168.178.91:6511/backend/webhooks/integrations/plex/nBrLZdk53g--24ryP
May 05, 2024 10:55:30.128 [124507096288056] DEBUG - Completed: [192.168.178.60:53778] 200 GET /:/timeline?ratingKey=10199&key=%2Flibrary%2Fmetadata%2F10199&playbackTime=0&playQueueItemID=9939&state=playing&hasMDE=1&context=home%3Ahub.continueWatching&row=0&col=0&time=0&duration=1776000&utm_referer=http%3A%2F%2F192.168.178.42%3A7575%2F (5 live) #f53be GZIP 11ms 453 bytes (pipelined: 5)
May 05, 2024 10:55:30.130 [124507070131000] DEBUG - [HttpClient/HCl#62a2] HTTP/1.1 (0.0s) 400 response from POST http://192.168.178.91:6511/backend/webhooks/integrations/plex/nBrLZdk53g--24ryP
May 05, 2024 10:55:30.131 [124506692152120] WARN - [Req#f53be] Webhook: Error delivering payload to http://192.168.178.91:6511/backend/webhooks/integrations/plex/nBrLZdk53g--24ryP: 400

Second error:
May 05, 2024 10:55:39.480 [124507070131000] DEBUG - [HttpClient/HCl#62a3] HTTP/1.1 (0.0s) 422 response from POST http://192.168.178.91:6511/backend/webhooks/integrations/plex/nBrLZdk53g--24ryP (reused)
May 05, 2024 10:55:39.480 [124506692152120] WARN - [Req#f5572] Webhook: Error delivering payload to http://192.168.178.91:6511/backend/webhooks/integrations/plex/nBrLZdk53g--24ryP: 422

I don't think I have to redact anything, so just posting the logs plainly. Basically, it's throwing a 400 and 422 error.
Also tried it with my public URL via the reverse proxy version of Ryot, same thing. Exact same errors.

**Comment by IgnisDa** [comment]
Does it post to `http://192.168.178.91:6511/backend/webhooks/integrations/plex/nBrLZdk53g--24ryP`? I think it should be `https://ryot.yourdomain.com/backend/webhooks/integrations/plex/nBrLZdk53g--24ryP`. Or am i missing something from your setup.

**Comment by IgnisDa** [comment]
> I am watching Curb your Enthusiasm on Plex. Next up is episode 9 of season 2. So I added Curb via the TV shows search to my library and, by marking all episodes as seen so the progress is in sync with Plex, put it "in Progress" (I can see the tag on top). Now I watch episode 9 on Plex. What I expect should happen is that it now marks episode 9 as seen on Ryot. But that doesn't happen.

I missed this sorry. Yes this is the expected behaviour.

**Comment by mwckr** [comment]
> Does it post to `http://192.168.178.91:6511/backend/webhooks/integrations/plex/nBrLZdk53g--24ryP`? I think it should be `https://ryot.yourdomain.com/backend/webhooks/integrations/plex/nBrLZdk53g--24ryP`. Or am i missing something from your setup.

When I enter the webhook from my internal IP it posts to that, when I swap it out with the webhook-URL from my proxy-URL, it posts to that. Both times same error, 400 and 422. I just tried both to see if I had a problem with my reverse proxy, because surely the servers should be able to communicate with each other on my local network, as all my other servers do.

**Comment by IgnisDa** [comment]
Can you look into the Ryot container logs? There should be some error logging when 422 is returned.

I added some more logging with the latest release (which should be published in an hour). Set env variable `RUST_LOG=ryot=debug` and paste it here. I would like to see if the body of the Plex webhook is correct.

**Comment by mwckr** [comment]
> Can you look into the Ryot container logs? There should be some error logging when 422 is returned.
> 
> I added some more logging with the latest release (which should be published in an hour). Set env variable `RUST_LOG=ryot=debug` and paste it here. I would like to see if the body of the Plex webhook is correct.

Will do. 
In the meantime, might I give one suggestion:

![grafik](https://github.com/IgnisDa/ryot/assets/96175020/ad09981d-9684-46a7-b9f4-43c02127e10a)

Would it be possible to get the option to swap these around? I'd like to see "current items" on top, and things that are upcoming below. Just a suggestion. Thanks.

**Comment by IgnisDa** [comment]
You can change the order in the dashboard preferences settings.

**Comment by mwckr** [comment]
> You can change the order in the dashboard preferences settings.

My god. How have I not seen that - I'm sorry. Thank you.

**Comment by mwckr** [comment]
I've found the issue in the logs, it works now.

The issue was my username. I didn't use the fullname because of technical reasons, so I tried the username. As plex doesn't use the username login anymore, instead relying on mail-adress, I used that. The logs told me "Ignoring non matching user", so I went back and looked at my account settings. I've not seen or used that username in years, so I didn't think of it.

I guess that's solved! Thanks for the very quick replies and help.
